### PR TITLE
[VersionControl] Fix problem detecting sln and csproj changes

### PR DESCRIPTION
**Pull request by jsuarezruiz**
In the StatusView, obtaining the status of files using `GetVersionInfo`, we obtain an incorrect value (without local changes). It seems to be a cache problem. For this reason, sometimes trying to add the change to the changeset we do not do it and we do not update the UI correctly.
Added validation where we verify the status of the file ignoring the cache (in necessary case).

Fixes VSTS #675483
Fixes VSTS #672672

### CODE REVIEW

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs [hunk 1]
```diff
@@ -1031,17 +1029,28 @@ bool OnFileStatusChanged (FileUpdateEventInfo args)
...
+				} else {
+					InalidateChangeSet (newInfo.LocalPath);
 				}
 			}
 			return true;
 		}
 
+		void InalidateChangeSet(FilePath path)
+		{
+			if (!changeSet.ContainsFile (path)) {
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs]: what happens if the changeset already contains the file?

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs [hunk 2]
```diff
@@ -999,6 +996,7 @@ bool OnFileStatusChanged (FileUpdateEventInfo args)
 			}
 
 			VersionInfo newInfo;
+
 			try {
 				// Reuse remote status from old version info
 				newInfo = vc.GetVersionInfo (args.FilePath);
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs]: Shouldn't this bypass the cache too?

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs]: Why would we bypass the cache? We shold have proper invalidation caused by FileService

Comment (sevoku) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs]: The main problem is that we update the cache in the background. On file changes we only mark an entry as invalid here: https://github.com/mono/monodevelop/blob/master/main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlFileSystemExtension.cs#L125 and then update them asynchronously here: https://github.com/mono/monodevelop/blob/master/main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs#L255.

So the right solution would be to fire `VersionControlService.NotifyFileStatusChanged` after the cache update (in `Repository.RunQueries()`) and not immediately on file change.

**Comment by sevoku** [comment]
@jsuarezruiz Lluis is right, see my comment here: https://github.com/mono/monodevelop/pull/7314#discussion_r263286534

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs (hunk 2 above)
Comment (slluis) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.Views/StatusView.cs]: @sevoku exactly. The repository knows when version status changes, so the repository should be responsible for notifying the change.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs [hunk 3]
```diff
@@ -23,181 +23,146 @@
...
-				fileStatus [versionInfo.LocalPath] = versionInfo;
-			} finally {
-				fileLock.ExitWriteLock ();
-			}
+			if (fileStatus.TryGetValue (versionInfo.LocalPath, out var vi) && vi.Equals (versionInfo)) {
+				vi.RequiresRefresh = false;
+				return;
+			}
+
+			fileStatus.AddOrUpdate (versionInfo.LocalPath, versionInfo, (f, v) => versionInfo);
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs]:
```suggestion
			fileStatus[versionInfo.LocalPath] = versionInfo;
```

The setter internally handles concurrent scenarios.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs [hunk 4]
```diff
@@ -23,181 +23,146 @@
...
+			foreach (var versionInfo in versionInfos) {
+				if (!versionInfo.IsInitialized)
+					versionInfo.Init (repo);
+
+				if (fileStatus.TryGetValue (versionInfo.LocalPath, out var vi) && vi.Equals (versionInfo)) {
+					vi.RequiresRefresh = false;
+					continue;
+				}
+
+				fileStatus.AddOrUpdate (versionInfo.LocalPath, versionInfo, (f, v) => versionInfo);
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs]:
```suggestion
			fileStatus[versionInfo.LocalPath] = versionInfo;
```

The setter internally handles concurrent scenarios.

**Comment by jsuarezruiz** [comment]
To simplify the review and the historical, we will unify this PR with https://github.com/mono/monodevelop/pull/7149 in a new one.

**Comment by jsuarezruiz** [comment]
New related PR: https://github.com/mono/monodevelop/pull/7345

**Pull request closed without merging**